Refactor and Expand Grouped Model Evaluation Plots with Residuals, Bootstrapped Curves, and Disparity Metrics

lshpaner opened the pull request:
## Summary

This is a comprehensive refactor and functional expansion of the grouped plotting utilities for model evaluation. This consolidates and standardizes core plotting logic across ROC, PR, calibration, residual, and disparity metric plots, with added support for subplots, bootstrapping, flexible color control, error bar overlays, and improved labeling/styling.

The overhaul improves **modularity**, **consistency**, and **flexibility**, with a focus on both **interpretability** and **reusability** for group-based fairness auditing and performance analysis.


## Refactor
- Unified plotting structure using `eq_plot_group_curves` and `eq_plot_residuals_by_group`.
- Extracted shared logic into helpers:
  - `get_layout`
  - `get_group_color_map`
  - `save_or_show_plot`
  - `plot_with_layout` (with support for `overlay_mode` to handle legend label formatting)
- Group-wise validation via `_filter_groups()` to filter out invalid groups (e.g., only one class in `y_true`) and support exclusion by name or minimum sample size.
- Bootstrap interpolation abstracted into `interpolate_bootstrapped_curves()` with ROC, PR, and calibration support.

---

## New Functionality

### 1. **Residuals Plotting**
- `eq_plot_residuals_by_group`:
  - Overlay, individual group, or subplot rendering
  - Group-colored scatterplots with optional centroid markers
  - Dynamic regression labels: R², MAE, residual mean, and `n`
  - Configurable transparency, color control, and subplot layout
  - Added `show_grid` parameter to toggle grid visibility

### 2. **Grouped Curves (ROC/PR/Calibration)**
- `eq_plot_group_curves`:
  - Supports `curve_type = 'roc' | 'pr' | 'calibration'`
  - Combined overlay or subplot view
  - Per-group line style config via `curve_kwgs`
  - Includes auto-reference lines (e.g., diagonal or prevalence)
  - Adaptive label verbosity via `label_mode` (full labels in overlay mode, simple labels in subplots/single-group mode)
  - Added `show_grid` parameter to toggle grid visibility
  - Custom legend positioning based on `curve_type` (e.g., `lower right` for ROC, `upper right` for PR)

### 3. **Bootstrapped Curves with Error Bars**
- `eq_plot_bootstrapped_group_curves`:
  - Shared x-grid interpolation via `interpolate_bootstrapped_curves()`
  - Plots mean curve with 95% confidence bands
  - Error bars at regular intervals (controlled by `bar_every`)
  - View modes: overlay, individual, or subplots
  - Curve metrics (AUC / AUCPR / Brier score) in legend with CI
  - Added `show_grid` parameter to toggle grid visibility

### 4. **Disparity Metrics Visualization**
- `eq_disparity_metrics_plot`:
  - Visualizes group-wise performance disparities from bootstrapped audits
  - Supports `violinplot` and `boxplot` styles via Seaborn
  - Color-coded categorical axis and stylized tick labels
  - Reference horizontal lines at y = 0, 1, 2
  - Supports legend toggling and custom layout configuration
  - Added `show_grid` parameter to toggle grid visibility

---

## Notable Improvements

- Consistent support for:
  - `figsize`, `dpi`, `n_rows`, `n_cols`, `save_path`, and `filename`
  - Tight grid layout control (`strict_layout`)
  - Group-level color control (`color_by_group`)
  - Explicit warnings for invalid data or missing group slices
  - Grid visibility toggling via `show_grid` parameter across all plotting functions

- Bootstrapping enhancements:
  - Handles missing/interrupted curve fits with NaN fallback
  - Uses `np.trapz` to compute and display AUC summary with CI
  - Optional vertical error bars overlaid at regular intervals

- Legend improvements:
  - Full labels in overlay mode (e.g., "AUC for asian = 0.47, Count: 244, Pos: 122, Neg: 122")
  - Simple labels in subplots/single-group mode (e.g., "AUC = 0.47")
  - Dynamic legend positioning based on `curve_type` in subplots/single-group mode

---

## File Structure

All functions are consolidated under a single plotting module (`plots.py`) and grouped as follows:

- **Shared Utilities**: layout, color mapping, saving, group filtering
- **Residual Plots**: `eq_plot_residuals_by_group`, `_plot_residuals_ax`
- **Grouped Curve Plots**: `eq_plot_group_curves`, `_plot_group_curve_ax`
- **Bootstrapped Curve Plots**: `eq_plot_bootstrapped_group_curves`, `_plot_bootstrapped_curve_ax`
- **Disparity Metrics**: `eq_disparity_metrics_plot`

---

##  Notes

- This PR supersedes legacy curve plotting utilities.
- Integrates with `regression_metrics()` and `EquiBoots` output structures.
- Dependencies: `matplotlib`, `seaborn`, `scikit-learn`, `scipy`, and `numpy`
- Calibration binning handled via bin-center interpolation
- updated suite of pytests

---

panas89 commented:
- Add a check to handle NaN values in [y_prob] and y_true
 - Add a check to ensure [y_true] exists for each group before filtering.
 - Add a check to ensure [n_rows * n_cols] is sufficient for the number of groups. (line 46)
 - Validate [curve_kwgs] before passing to plot_group_curve_ax if it has the correct arguments
 

Looks great, some checks to be run while plotting, ensuring the quality of input data or parameters can approve after they have been added. Tests pass!

panas89 commented:
line 536 np.trapz depreceated maybe change to np.trapezoid

lshpaner commented:
> line 536 np.trapz depreceated maybe change to np.trapezoid

It's an alias introduced in `numpy===1.24`; should still work w/ `np.trapz`

lshpaner commented:
> * Add a check to handle NaN values in [y_prob] and y_true
> * Add a check to ensure [y_true] exists for each group before filtering.
> * Add a check to ensure [n_rows * n_cols] is sufficient for the number of groups. (line 46)
> * Validate [curve_kwgs] before passing to plot_group_curve_ax if it has the correct arguments
> 
> Looks great, some checks to be run while plotting, ensuring the quality of input data or parameters can approve after they have been added. Tests pass!

----

**Update: Added Requested Checks**

- Added a check to handle `NaN` values in `y_prob` and `y_true` using a new helper `_validate_plot_data`, which also validates the existence of `y_true` and `y_prob`/`y_pred` for each group before filtering.
- Added a check in `plot_with_layout` to ensure `n_rows * n_cols` is sufficient for the number of groups in subplots mode, raising a `ValueError` if the grid is too small.
- Added validation for `curve_kwgs` in `eq_plot_group_curves` using `_validate_plot_kwargs` to ensure it contains valid `Matplotlib` plot arguments. Also extended this validation to `line_kwgs` in both `eq_plot_group_curves` and `eq_plot_bootstrapped_group_curves`.
- New notebook added to additionally validate these, called `test_plot_exceptions.ipynb`
- Updated unittests; coverage increased to 91%:

  ```text
  ---------- coverage: platform linux, python 3.11.0-final-0 -----------
  Name                              Stmts   Miss  Cover
  -----------------------------------------------------
  src/equiboots/EquiBootsClass.py     157     25    84%
  src/equiboots/__init__.py            18      0   100%
  src/equiboots/logo.py                 1      0   100%
  src/equiboots/metrics.py             72      0   100%
  src/equiboots/plots.py              310     28    91%
  -----------------------------------------------------
  TOTAL                               558     53    91%
  ```

panas89 commented:
tested, test pass!
